### ISSUE TITLE
[FasterTransformer/V2] No speedUp when the sequence length is large

### ISSUE THREAD

**Issue by ligonzheng** [NVIDIA/DeepLearningExamples#562]
Hello,
    I had implemented a gpt2 model according to FT. Then I tested the performance between pytorch(Fairseq) and FT, This is the result : 

Setting : batch = 1, hidden_units = 1024, head_num = 16, size_per_head = 64
time : ms
seq_len : 8  16  32  64  128  256   512    800
pytorch: 21  23  23   23  28   23.6  22.6   24
___FT : 6.2 6.4 6.7 7.6  8.6  12.3   24     34.7
 
From this table , it seems like the FT is much worse than pytorch when the seq_len is larger than about 500.  it's very unacceptable that the FT is slower.  And why the performance of pytorch is almost never change with the increase of sequence length ? 
It's also has the same phenomenon when I test the masked BERT model.
Anyone help ?

**Comment by byshiue** [comment]
I think you use the decoding op. If yes, this is very strange because no matter which implementation, the inference time should be proportional to the sequence length. But in your case, both pytorch and FT are not. 
Have you verified the correctness by trying translate some benchmark?

**Comment by ligonzheng** [comment]
Yes, I use the decoding op in this test. And I also had verified the correctness.  The max relative difference is 10e-3 and the max absolute difference is 10e-4. 
In my case , The FT result is proportional to the sequence length , Why you said both pytorch and FT are not ?
I had check the Readme of FT, then it shows that the tensorflow and the FT are proportional to sequence length . This is really strange that the pytorch are not in this rule.

**Comment by byshiue** [comment]
I am not sure why pytorch not.
But even for FT, assume the inference time of sequence length 8 is 6 ms, then the time of sequence length 16 should be larger than 12 ms because the time increase super-linearly. But in your cases, the inference time of sequence 8, 16, 32 are similar, this is strange.

**Comment by ligonzheng** [comment]
Hello, 
I used nvvp to analysis the reason. I find a very interesting point . A memory copy from device to host occupied too much time, but the actually time it needed is very little.
![Screenshot from 2020-06-19 17-46-37](https://user-images.githubusercontent.com/33247341/85120029-39359a80-b255-11ea-97b9-5bb37a8b04d1.png)
See this picture, There are two duration at line 3 and line 8. The really needed time is 1.216us, but the time it consume is 11.48906ms.
Do you know why ?

**Comment by ligonzheng** [comment]
Hello, 
I used nvvp to analysis and there is a  interesting thing.
![Screenshot from 2020-06-19 17-46-37](https://user-images.githubusercontent.com/33247341/85245202-70849100-b479-11ea-87b1-a828b612a6bf.png)
As this picture shows. The cudaMemcpy consume too much time, but the actually time it needed is very small.
I remove the cudaMemcpy ,then the performance is up to 10x.

**Comment by byshiue** [comment]
> Hello,
> I used nvvp to analysis and there is a interesting thing.
> ![Screenshot from 2020-06-19 17-46-37](https://user-images.githubusercontent.com/33247341/85245202-70849100-b479-11ea-87b1-a828b612a6bf.png)
> As this picture shows. The cudaMemcpy consume too much time, but the actually time it needed is very small.
> I remove the cudaMemcpy ,then the performance is up to 10x.
> I also tried to use pinned memory , but no increment for performance.
> 
> For the total process, this kind of bad cudaMemcpy occupied too much time. As the second picture shows :
> ![Screenshot from 2020-06-22 11-16-51](https://user-images.githubusercontent.com/33247341/85245415-0f10f200-b47a-11ea-98dd-faa44cdf8862.png)

So, do you have any other problem?

**Comment by ligonzheng** [comment]
Yes, The key is the time of cudaMemcpy.  the actually consume time is much longer than it need. 
Total time for it consumed is 10000x than it copied time.  I want to know why and could give me some advise to avoid it ?

**Comment by byshiue** [comment]
I cannot know the information from your graph.
From the first graph, the cudaMemcpy only takes 1.2 us, a very short time. 
From the second graph, I also see that most time on consumed on computing, but not memory copy. So, I cannot understand why you say that cudaMemcpy consumes most of time. 
Please give more information.

For the decoding, the FasterTransformer requires 1 memory copy after decoding one word, so there are "sequence_length" times of memory copy for one inference. But the data size is very small, and hence the copy time is very short.

**Comment by ligonzheng** [comment]
From the first graph, there are two duration ! The first is the actually time it consumed ,the second time is the really copy time it used. There is a big gap between this two duration. That's my doubt. 
I know the decoding only need 1 memory copy after decoding one word, but I remake it for some reason. I copy the last decoder's result  to CPU .

**Comment by byshiue** [comment]
From your second graph, I guess that the real copy time is only 1 us; 12 ms means that the time of launching kernel. 
Another question is, the time of memory copy is one step or one inference? If one step, then the inference time of 800 sequence length is strange. In fact, decoding 800 words only requires 34 ms is also strange. 
Can you provide the kernel statistic of sequence length 32 and 800 by nvvp? 
You can use "nvvp ./bin/your_program <your_arguments_1> ... " or "nvvp python your_program" to get the information directly.

**Comment by ligonzheng** [comment]
I do not think  12ms means the time of kernel launching. On the one hand, kernel launch may not need so much time. On the other hand, some copy from host to device or device to device does not need so much time. 
The second question, the time of memory copy is one step.  I just count the time of my N decoders forward and the data copy between host and gpu.  34ms is also the time of one step.

**Comment by byshiue** [comment]
I have misunderstood many things. 
From your nvvp files, I understand what you means. 
1. Why the memory copy consumes lots of time? 
Ans: For memory copy, the "Runtime API" means the kernel launch time, but not the real copy time. The real copy time is under "Geforce GTX 1080 Ti -> Context 1 -> MemCpy (xxx)". I guess that the "cudaMemcpy" runtime API consumes lots of time is because the cpu needs to waits the next kernel to launch. In FT, the cpu will launch the kernels without waiting the gpu complete the jobs. But in pytorch, it may be not.

2. Why the FT is slower than pytorch when sequence length larger?
Ans: The reason is, the masked_attention_kernel is implemented by cuda kernels. This implementation is more efficient when the max sequence length is small, because the kernel launch time is close to the execution time in such case (10 us and 18 us). So, fusing many operations into one kernel is more efficient. 
However, when the sequence length is large enough, using cuda kernel to implement the batched gemm multiplication is not a good option.

**Comment by byshiue** [comment]
Another question is: if the prof_ft_1000.nvvp is the profiling file for sequence length 1000, the inference time of each step is only 20 ms, which is also shorter than your demo above.

**Comment by ligonzheng** [comment]
Really thank you for your explanation. I understand why the memory copy consumes lots of time. 
For the another question , My demo above is the data that run on 2080ti but the nvvp files is running on 1080ti. For some reason I can not execute the nvvp on the 2080ti.
So, what's the mainly meathod for FT ? the fuse of the kernels so that reduce the times of launch ? the usage of shared memory ? the different algo choices for cublas gemm ?
Is there a way to improve the performance of large sequence ? Even if the FT has the same performance as pytorch version.

**Comment by byshiue** [comment]
1. The optimization methods are like you say: the fuse of the kernels so that reduce the times of launch ? the usage of shared memory ? the different algo choices for cublas gemm. 
    But they have different effects. Fusing kernels reduce the time of launching kernel, which brings benefit when the batch size is small. Shared memory improves the efficiency of custom kernels. choosing gemm algorithm provides the best performance no matter the shape, and brings benefit when the batch size is large. 

2. It is very hard to further optimize the FT. Some possible direction is: (1) removing the padding like effective transformer. This method is useful for the encoder case, and would be added into FT in the next update. (2) INT 8 quantization. But there are lots of works to use the INT8 quantization. (3) Further optimization for the custom kernels. But I am not sure which kernel can be further optimized. 

3. If you cannot use the nvvp, you can try to use [nsight systems](https://developer.nvidia.com/nsight-systems), which is the next version nvvp. The overhead of profiling of nsight systems is smaller than nvvp.

**Comment by venuswu** [comment]
> I have misunderstood many things.
> From your nvvp files, I understand what you means.
> 
> 1. Why the memory copy consumes lots of time?
>    Ans: For memory copy, the "Runtime API" means the kernel launch time, but not the real copy time. The real copy time is under "Geforce GTX 1080 Ti -> Context 1 -> MemCpy (xxx)". I guess that the "cudaMemcpy" runtime API consumes lots of time is because the cpu needs to waits the next kernel to launch. In FT, the cpu will launch the kernels without waiting the gpu complete the jobs. But in pytorch, it may be not.
> 2. Why the FT is slower than pytorch when sequence length larger?
>    Ans: The reason is, the masked_attention_kernel is implemented by cuda kernels. This implementation is more efficient when the max sequence length is small, because the kernel launch time is close to the execution time in such case (10 us and 18 us). So, fusing many operations into one kernel is more efficient.
>    However, when the sequence length is large enough, using cuda kernel to implement the batched gemm multiplication is not a good option.

**why In FT, the cpu  launch the kernels without waiting the gpu**

**Comment by byshiue** [comment]
In FT, we do not sync all device after each inference. This will save few time on kernel launching because the cpu can launch the kernels if it is idle. 
In pytorch, it is often to sync all device after each inference automatically. 
In small batch size, the performance of them may be little different. 
In large batch size, there is no effect.

**Comment by byshiue** [comment]
Close this bug because it is inactivated. Feel free to re-open this bug if you still have any problem.
